Update cherrypick doc for release notes to more accurately reflect current process.

david-mcmahon opened the pull request:
## Contributor guidelines
1. Please read our [contributor guidelines](https://github.com/kubernetes/kubernetes/blob/master/CONTRIBUTING.md).
2. See our [developer guide](https://github.com/kubernetes/kubernetes/blob/master/docs/devel/development.md).
3. Follow the instructions for [labeling and writing a release note for this PR](https://github.com/kubernetes/kubernetes/blob/master/docs/devel/pull-requests.md#release-notes) in the block below.

eparis commented:
@bgrant0607 @zmerlynn @roberthbailey Personally I think `release-note-note` is ok on some cherry-picks. Things which have NO user impact. Jenkins changes, OWNERS, etc.  But it should be exceptional, not common....

david-mcmahon commented:
@bgrant0607 what say you?  We had discussed this at one point and you had mentioned wanting to keep this hard rule on cherrypicked PRs.  I see the point of wanting to do so but I also agree with @eparis that there are cases where none is also appropriate.  I lean toward the more flexible option of allowing release-note-none in rare cases and we can update this doc to reflect the preference.

eparis commented:
Today the bot enforces that the parent PR have `release-note` or `release-note-action-required`. If the parent does not have one of those two (or a parent cannot be located) the bot will require the cherrypick PR to have a label. If the parent does not have one of those two labels the cherrypick PR will get a comment like the following:

https://github.com/kubernetes/kubernetes/pull/24869#issuecomment-215146846

In which case the person who owns the branch will need to look at it. Either adding `release-note-none` to the cherrypick PR or fixing the parent PR to include one of the two above named labels.

I personally like having a backstop against `release-note-none` on the parent. And think, as is the case with #24869 which only touches e2e, no release note makes sense. No one downloading and running 1.2.x cares a whit if we changed the e2e test...

bgrant0607 commented:
I'm fine with @eparis's proposal

roberthbailey commented:
I second @eparis's proposal.

zmerlynn commented:
I'm fine with it, too.

david-mcmahon commented:
SGTM. PTAL.

eparis commented:
That's not quite what's happening today.
- if the parent has `release-note` or `release-note-action-required` the cherrypick needs nothing.
- if no parent can be determined the cherrypick must follow the process.
- If the parent has `release-note-none` or didn't follow the process the cherrypick must follow the process

eparis commented:
we could automate copying `release-note` and `release-note-action-required` from the parent to the cherrypick, and then force all cherrypicks to follow the process...

david-mcmahon commented:
@bgrant0607 had originally wanted to keep things simple by keeping most or all accounting on the original/parent PR.  I liked that idea in theory but in practice there's quite a bit happening on the cherry picked PRs too so maybe just sweeping over the release-note (and milestone?) bits would work best.  I dislike the amount of churn here.  It's confusing to contributors (and even we implementers), but I also don't want that to stop us from doing what's best.

david-mcmahon commented:
@eparis, PTAL.  The doc addresses your points 1 and 3.  Point 2 is the default behavior of the standard [pull-request.md](https://github.com/kubernetes/kubernetes/blob/master/docs/devel/pull-requests.md#release-notes) doc (linked from this doc as well).

david-mcmahon committed:
Update pull request and cherrypick docs for release notes to more accurately reflect current process.

david-mcmahon commented:
ping.

david-mcmahon commented:
@zmerlynn @eparis @roberthbailey anyone, LGTY?

roberthbailey commented on docs/devel/cherry-picks.md:
```diff
@@ -42,12 +42,13 @@ depending on the point in the release cycle.
 label changes to the **master** branch PR:
-  * Remove release-note-label-needed
-  * Add an appropriate release-note-(!label-needed) label
+  * Remove the `release-note-label-needed` label
```
This line can be removed, since it applies to all PRs against master and you don't have to do anything special for a cherry pick candidate.

david-mcmahon replied:
You are correct.  That shouldn't be there.  Gone.

roberthbailey commented on docs/devel/pull-requests.md:
```diff
@@ -86,7 +86,7 @@ This section applies only to pull requests on the master branch.
 The only exception to these rules is when a PR is not a cherry-pick and is
 targeted directly to the non-master branch.  In this case, a `release-note-*`
-label is required for that non-master PR.
+label (that is not release-note-none) is required for that non-master PR.
```
Didn't we agree that there were cherry pick PRs (e.g. to fix test infra) that we'd take onto the release branch without a release note?

david-mcmahon replied:
You are correct. That shouldn't be there. Gone.

david-mcmahon commented:
rebase done.  changes done.  Thanks @roberthbailey for looking at this.

david-mcmahon commented:
@k8s-bot retest: UNKNOWN

david-mcmahon commented:
@eparis this is more concise now.  I removed all of the bits about what should already be on the master PR.  It just mentions the few labels needed now with the clarification on how release-note-none is handled.  Does this look right to you?

brendandburns commented on docs/devel/cherry-picks.md:
```diff
@@ -42,14 +42,12 @@ depending on the point in the release cycle.
-  * Add an appropriate milestone
-  * Add the `cherrypick-candidate` label
+1. To get a PR merged to the release branch, ensure the following labels are
+   on the **master** branch PR:
```
Clarify this wording a little bit? 

"The PR containing the commits that will be merged into the master branch"?

david-mcmahon replied:
I've updated it this part a bit, but I'm not seeing where you're going with that.

brendandburns commented on docs/devel/cherry-picks.md:
```diff
@@ -42,14 +42,12 @@ depending on the point in the release cycle.
-  * Add the `cherrypick-candidate` label
+1. To get a PR merged to the release branch, ensure the following labels are
+   on the **master** branch PR:
+  * An appropriate milestone
```
... (e.g. v1.3)

david-mcmahon replied:
Done.
